### PR TITLE
Extend head_test.go and db_test.go with OOO histogram test cases

### PR DESCRIPTION
**Pull request by fionaliao**
- Updated OOO tests in head_test.go with histogram test cases
- Added processing of histograms to the WBL
- Updated a few OOO tests in db_test.go as well
- Probably could do a bit more refactoring but I don't think that needs to be done before this is merged

### CODE REVIEW

**Comment by fionaliao** [comment]
I'll finish https://github.com/prometheus/prometheus/issues/11329#issuecomment-1702815572 before merging this PR to avoid the merge conflicts that would otherwise appear

**Comment by fionaliao** [comment]
WBL changes done, have cherry-picked those commits to this branch, so now ready to review

**Comment by fionaliao** [comment]
Two of the commits were cherry picked from the main prometheus branch, so probably easier to just review the other commits: https://github.com/carrieedwards/prometheus/pull/1/files/ede1f2ef258bf799a978e0dc843febe11dceb937..533b3913c50f3c77da76be8b158f46289e3f6549

File: tsdb/head_test.go [hunk 1]
```diff
@@ -5219,3 +5351,45 @@ func TestSnapshotAheadOfWALError(t *testing.T) {
...
 	require.NoError(t, head.Close())
 }
+
+type sampleTypeScenario struct {
+	sampleType string
+	appendFunc func(appender storage.Appender, lbls labels.Labels, ts int64, value int64) (storage.SeriesRef, error, sample)
+	sampleFunc func(ts, value int64) sample
+}
+
+var sampleTypeScenarios = map[string]sampleTypeScenario{
```
Comment (carrieedwards) [inline tsdb/head_test.go:5361]: I like this refactoring so that we can re-use the same appendFunc and sampleFunc for the different data types

**Review by carrieedwards — APPROVED** [review]
LGTM. Thank you for the code deduplication!

**Pull request merged**